aws-storage-s3/src/androidTest/java/com/amplifyframework/storage/s3/AWSS3StorageUploadTest.java

```diff
@@ -0,0 +1,224 @@
+
+    private static final long LARGE_FILE_SIZE = 10 * 1024 * 1024L; // 10 MB
+    private static final long SMALL_FILE_SIZE = 100L;
+
+    private final String largeFileName = "large-test-" + System.currentTimeMillis();
+    private final String smallFileName = "small-test-" + System.currentTimeMillis();
+
+    private File largeFile;
+    private File smallFile;
+
+    private StorageUploadFileOptions options;
+    private String largeFileKey;
+    private String smallFileKey;
+
+    /**
+     * Sets up the required files for testing transfers.
+     * @throws Exception if fails to create temp files
+     */
+    @Before
+    public void setUp() throws Exception {
+        // Randomly write objects to upload
+        largeFile = createTempFile(largeFileName, LARGE_FILE_SIZE);
+        smallFile = createTempFile(smallFileName, SMALL_FILE_SIZE);
+
+        // Always interact with PUBLIC access for consistency
+        options = StorageUploadFileOptions.builder()
+                .accessLevel(DEFAULT_ACCESS_LEVEL)
+                .build();
+
+        largeFileKey = getS3Key(DEFAULT_ACCESS_LEVEL, largeFileName);
+        smallFileKey = getS3Key(DEFAULT_ACCESS_LEVEL, smallFileName);
+    }
+
+    /**
+     * Cleans up the S3 bucket for files that were uploaded
+     * during testing processes.
+     */
+    @After
+    public void cleanUp() {
+        cleanUpS3Object(largeFileKey);
+        cleanUpS3Object(smallFileKey);
+    }
+
+    /**
+     * Tests that small file (single-part) uploads successfully.
+     *
+     * @throws StorageException if upload fails
+     */
+    @Test
+    public void testUploadSmallFile() throws StorageException {
+        StorageUploadFileResult result = Await.<StorageUploadFileResult, StorageException>result(
+                (onResult, onError) ->
+                        Amplify.Storage.uploadFile(
+                                smallFileName,
+                                smallFile.getAbsolutePath(),
+                                onResult,
+                                onError
+                        )
+        );
+        assertEquals(smallFileName, result.getKey());
+        assertS3ObjectExists(smallFileKey);
+    }
+
+    /**
+     * Tests that large file (multi-part) uploads successfully.
+     *
+     * @throws Exception if upload fails
+     */
+    @Test
+    public void testUploadLargeFile() throws Exception {
+        final CountDownLatch completed = new CountDownLatch(1);
+        Amplify.Storage.uploadFile(
+                largeFileName,
+                largeFile.getAbsolutePath(),
+                onResult -> {
+                    assertEquals(largeFileName, onResult.getKey());
+                    assertS3ObjectExists(largeFileKey);
+                    completed.countDown();
+                },
+                onError -> fail("Upload was not successful.")
+        );
+        completed.await(DEFAULT_TIMEOUT_IN_SECONDS, TimeUnit.SECONDS);
```

Comment:
One improvement that could be made here is that rather than have to wait for the timeout if it fails you have a way of immediately returning.

Reply:
Wouldn't the test immediately terminate upon encountering `fail("Upload was not successful.")`?

Reply:
No because the completed latch won't count down in that case and the test waits for the latch to either countdown or timeout.

Reply:
Oooh, sorry I'm an idiot - I missed that fail was a function not a log statement. Never mind.